pdns/lua-record.cc

```diff
@@ -108,25 +211,27 @@ class IsUpOracle
   {
     setStatus(cd, true);
   }
-
-  bool upStatus(const ComboAddress& rem, const std::string& url=std::string(), const opts_t& opts = opts_t())
-  {
-    CheckDesc cd{rem, url, opts};
-    std::lock_guard<std::mutex> l(d_mutex);
-    return d_statuses[cd].status;
-  }
 };
 
 bool IsUpOracle::isUp(const CheckDesc& cd)
 {
-  std::lock_guard<std::mutex> l(d_mutex);
-  auto iter = d_statuses.find(cd);
-  if(iter == d_statuses.end()) {
-    d_statuses[cd]=Checker{std::thread(&IsUpOracle::checkTCPThread, this, cd.rem, cd.opts), false};
-    return false;
+  if (!d_checkerThread) {
+    d_checkerThread = std::unique_ptr<std::thread>(new std::thread(&IsUpOracle::checkThread, this));
   }
-  return iter->second.status;
-
+  time_t now = time(nullptr);
+  {
+    ReadLock lock{&d_lock};
+    auto iter = d_statuses.find(cd);
+    if (iter != d_statuses.end()) {
+      iter->second->lastAccess = now;
+      return iter->second->status;
+    }
+  }
+  {
+    WriteLock lock{&d_lock};
+    d_statuses[cd] = std::unique_ptr<CheckState>(new CheckState{now});
```

Comment:
I'm not sure if we care but it looks like another thread might be able to insert the same entry between the time we release the lock then acquire it again.